src/osp/Active/CommonPhysics.cpp

```diff
@@ -0,0 +1,148 @@
+/**
+ * Open Space Program
+ * Copyright � 2019-2020 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "CommonPhysics.h"
+
+#include <iostream>
+
+namespace osp::phys
+{
+
+float col_shape_volume(ECollisionShape shape, Vector3 scale)
+{
+    constexpr float pi = 3.14159265359f;
+    switch (shape)
+    {
+    case ECollisionShape::NONE:
+        return 0.0f;
+    case ECollisionShape::SPHERE:
+        // Default radius: 1
+        return (4.0f / 3.0f) * pi * scale.x() * scale.x() * scale.x();
+    case ECollisionShape::BOX:
+        // Default width: 2
+        return 2.0f*scale.x() * 2.0f*scale.y() * 2.0f*scale.z();
+    case ECollisionShape::CYLINDER:
+        // Default radius: 1, default height: 2
+        return pi * scale.x() * scale.x() * 2.0f*scale.z();
+    case ECollisionShape::CAPSULE:
+    case ECollisionShape::CONVEX_HULL:
+    case ECollisionShape::TERRAIN:
+    case ECollisionShape::COMBINED:
+    default:
+        std::cout << "Error: unsupported shape for volume calc\n";
+        return 0.0f;
+    }
+}
+
+Matrix3 transform_inertia_tensor(Matrix3 I, float mass, Vector3 translation, Matrix3 rotation)
+{
+    // Apply rotation via similarity transformation
+    I = rotation.transposed() * I * rotation;
+
+    // Translate via tensor generalized parallel axis theorem
+    using Magnum::Math::dot;
+    Vector3 r = translation;
+    Matrix3 outerProductR = {r * r.x(), r * r.y(), r * r.z()};
+    Matrix3 E3 = Matrix3{};
+
+    return I + mass * (dot(r, r) * E3 - outerProductR);
```

Comment:
Weird parents?

Reply:
The parentheses? They're as intended

Reply:
Ah, so you're relying on the operator precedence rules. Ok.